altair/vegalite/v5/api.py

```diff
@@ -913,10 +933,23 @@ def condition(
 # Top-level objects
 
 
+def _top_schema_base(obj: Any, /):  # -> <subclass of SchemaBase and TopLevelMixin>
+    """Enforces an intersection type w/ `SchemaBase` & `TopLevelMixin` objects.
+
+    Use for instance methods.
+    """
+    if isinstance(obj, core.SchemaBase) and isinstance(obj, TopLevelMixin):
+        return obj
+    else:
+        msg = f"{type(obj).__name__!r} does not derive from {type(core.SchemaBase).__name__!r}"
+        raise TypeError(msg)
+
+
 class TopLevelMixin(mixins.ConfigMethodMixin):
     """Mixin for top-level chart objects such as Chart, LayeredChart, etc."""
 
     _class_is_valid_at_instantiation: bool = False
+    data: Any
```

Comment:
Fixes all of these, with no runtime changes:
https://github.com/vega/altair/blob/9afd374a83f954e79a5fd6e4f8a8a9c4e5381c04/altair/vegalite/v5/api.py#L2964
https://github.com/vega/altair/blob/9afd374a83f954e79a5fd6e4f8a8a9c4e5381c04/altair/vegalite/v5/api.py#L2870
https://github.com/vega/altair/blob/9afd374a83f954e79a5fd6e4f8a8a9c4e5381c04/altair/vegalite/v5/api.py#L2777
https://github.com/vega/altair/blob/9afd374a83f954e79a5fd6e4f8a8a9c4e5381c04/altair/vegalite/v5/api.py#L2787